src/oidc/tokenRefresher.ts

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2023 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { IdTokenClaims, OidcClient, WebStorageStateStore } from "oidc-client-ts";
+
+import { TokenRefreshFunction } from "../http-api";
+import { IDelegatedAuthConfig } from "../client";
+import { generateScope } from "./authorize";
+import { discoverAndValidateAuthenticationConfig } from "./discovery";
+import { logger } from "../logger";
+
+/**
+ * @experimental
+ * Class responsible to refreshing OIDC access tokens
+ */
+export abstract class OidcTokenRefresher {
```

Comment:
It would be nice to explain to client implementors how this should be used. Given that it is an abstract class, the docs should probably recognise that. Something like:

```suggestion
/**
 * Base class for refreshing OIDC access tokens.
 *
 * Client implementations will typically just need to provide an implementation of {@link #persistTokens}.
 * You can then use {@link #doRefreshAccessToken} in {@link ICreateClientOpts.tokenRefreshFunction}.
 */
export abstract class OidcTokenRefresher {
```

Reply:
Will add the link to `ICreateClientOpts` once `tokenRefreshFunction` is added there